sql/src/main/java/org/apache/druid/sql/calcite/schema/DruidSchema.java

```diff
@@ -103,9 +104,10 @@
   // Protects access to segmentSignatures, mutableSegments, segmentsNeedingRefresh, lastRefresh, isServerViewInitialized
   private final Object lock = new Object();
 
-  // DataSource -> Segment -> RowSignature for that segment.
-  // Use TreeMap for segments so they are merged in deterministic order, from older to newer.
-  private final Map<String, TreeMap<DataSegment, RowSignature>> segmentSignatures = new HashMap<>();
+  // DataSource -> Segment -> SegmentMetadataHolder(contains RowSignature) for that segment.
```

Comment:
Looks that this is still needed, but I think the current concurrency control mechanism is too complicated. It's spanned over several classes including `DruidSchema`, `SystemSchema`, and `SegmentMetadataHolder`. This can introduce potential bugs and make maintenance difficult. Can we simplify this?

Reply:
Yes this lock is required. And I have tried to keep the data safe, do you see any safety issues ? How do you suggest to simplify ?

Reply:
It depends on how latest result you want vs how simple you want to make this. If you want the very latest segment information, the very recent changes should be reflected to the query result. To do so, everything should be mutable and lock should be acquired before updating mutable objects. However, I don't think this is the case. System schema doesn't change so quickly like hundreds times in a second. 

I think it's much better to make `SegmentMetadataHolder` immutable and `segmentMetadataInfo` as a normal map. `getSegmentMetadata()` should return an immutable map holding the snapshot of `SegmentMetadataHolder` under the `lock`.

Reply:
Made `SegmentMetadataHolder` immutable, and it seems `segmentMetadataInfo` can be a HashMap now protected by `lock`. Yes `getSegmentMetadata()` is also returning `Map<DataSegment, SegmentMetadataHolder>` created under the `lock`

Reply:
Thanks.